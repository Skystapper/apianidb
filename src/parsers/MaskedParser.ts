import { MapField } from './field_types/'
import type { AnidbCacheImplType } from '../utils/AnidbCache'

export type MaskDefinition<T> = Record<T & string, readonly [number, number, (...args: Array<any>) => any]>;

export default class MaskUtil<T extends string> {
  fields: any[]
  mask: string
  static readonly mask_definition: MaskDefinition<string>
  parseData: (data: string) => Record<string, any>

  constructor (mask: Array<T> | string) {
    if (typeof mask === 'string') {
      this.fields = this._decode(mask)
      this.mask = mask
    } else {
      this.fields = this._prepareMaskArray(mask)
      this.mask = this._encode(this.fields)
    }
    const { mask_definition } = this.constructor as typeof MaskUtil
    this.parseData = MapField(new Map(this.fields.map(field => [ field, mask_definition[field][2] ])))
  }

  verifyObject = (object: Awaited<ReturnType<AnidbCacheImplType['get']>>) => {
    return !!(object && typeof object === 'object' && this.fields.every(f => f in object ))
  }

  _prepareMaskArray (mask_array: Array<T>) {
    const deduplicated = Array.from(new Set(mask_array))
    const {
      mask_definition,
    } = this.constructor as typeof MaskUtil
    // check all exist!
    let incorrect_field = ''

    if (!deduplicated.every(i => {
      incorrect_field = i
      return i in mask_definition // .hasOwnProperty(i)
    })) {
      throw new Error(`Bad field: ${ incorrect_field }`)
    }

    return deduplicated.sort((a, b) => {
      const a_value = mask_definition[a][0] * 1000 - mask_definition[a][1]
      const b_value = mask_definition[b][0] * 1000 - mask_definition[b][1]
      return a_value - b_value
    })
  }

  // TODO: maybe switch to buffer?
  _encode (mask_array: Array<T>): string {
    const {
      mask_definition,
    } = this.constructor as typeof MaskUtil
    const mask = Array(Object.values(mask_definition).reduce((m, c) => c[0] + 1 > m ? c[0] + 1 : m, 0)).fill(0)

    // const mask = [ 0, 0, 0, 0, 0, 0, 0 ]
    for (const mask_item of mask_array) {
      const flag_definition = mask_definition[mask_item]
      const [ byte_num, bit_value ] = flag_definition
      mask[byte_num] = (mask[byte_num] || 0) + bit_value
    }

    const buf = Buffer.from(mask)
    return buf.toString('hex')
  }

  // Shitty, but only used in debugging anyway
  _decode (encoded_mask: string): Array<any> {
    const {
      mask_definition,
    } = this.constructor as typeof MaskUtil
    const mask_array = []

    const enc_buf = Buffer.from(encoded_mask.padEnd(14, '0'), 'hex')

    for (let i = 0; i < 7; i++) {
      const relevant_keys = Object.keys(mask_definition).filter(m => mask_definition[m][0] === i)
      const value = enc_buf.readInt8(i)

      for (const key of relevant_keys) {
        const flag_v = mask_definition[key][1]

        if ((value & flag_v) > 0) {
          mask_array.push(key)
        }
      }
    }

    return mask_array
  }

}