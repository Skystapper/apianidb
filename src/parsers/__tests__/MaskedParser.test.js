import MaskedParser from '../MaskedParser'

class TestMaskedParser extends MaskedParser {
  static mask_definition = {
    field_1: [ 0, 64, String ],
    field_2: [ 0, 32, String ],
    field_3: [ 0, 16, String ],
    field_4: [ 0, 8, String ],
    field_5: [ 0, 4, String ],
    field_6: [ 0, 2, String ],
    field_7: [ 0, 1, String ],

    field_8: [ 1, 64, String ],
    field_9: [ 2, 64, String ],
  }

}

describe('new MaskedParser()', () => {

  it('new MaskedParser(array[])', () => {
    const maskedParser = new TestMaskedParser([ 'field_8', 'field_2', 'field_1', 'field_4', 'field_3', 'field_7', 'field_9', 'field_2' ])
    expect(maskedParser.mask).toBe('794040')
    expect(maskedParser.fields).toMatchSnapshot()
  })

  it('bad field', () => {
    expect(
      () => new TestMaskedParser([ 'bad_field' ])
    ).toThrowError(new Error('Bad field: bad_field'))
  })

  it('new MaskedParser(string)', () => {
    const maskedParser = new TestMaskedParser('794040')
    expect(maskedParser.mask).toBe('794040')
    expect(maskedParser.fields).toMatchSnapshot()
  })
})

describe('verifyObject', () => {
  const maskedParser = new TestMaskedParser([ 'field_1', 'field_2' ])
  it('good object', () => {
    maskedParser.verifyObject({ field_1: 'field_1', field_2: 'field_2' })
  })
  it('bad object', () => {
    maskedParser.verifyObject({ field_1: 'field_1' })
  })
  it('no object', () => {
    maskedParser.verifyObject(undefined)
  })

})