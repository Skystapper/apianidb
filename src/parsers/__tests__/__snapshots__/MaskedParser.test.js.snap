// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new MaskedParser() new MaskedParser(array[]) 1`] = `
[
  "field_1",
  "field_2",
  "field_3",
  "field_4",
  "field_7",
  "field_8",
  "field_9",
]
`;

exports[`new MaskedParser() new MaskedParser(string) 1`] = `
[
  "field_1",
  "field_2",
  "field_3",
  "field_4",
  "field_7",
  "field_8",
  "field_9",
]
`;
